drivers: virtio: add VIRTIO MMIO transport driver

soburi opened the pull request:
Adds a virtio-mmio driver that is based on the virtio-pci design from https://github.com/zephyrproject-rtos/zephyr/pull/83892.

This is worked with virtio-rng driver
https://github.com/zephyrproject-rtos/zephyr/pull/89439

And it works with the same API as virtio-pci.

soburi committed:
drivers: virtio: add VIRTIO MMIO transport driver

Add virtio-mmio driver

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
boards: qemu: cortex_a53: add virtio-mmio configuration

Add `virtio-mmio` nodes that are retrieved from qemu internal
devicetree dump, to dts.

Also, enable the virtio-rng device on the command line.

Note: Ideally, it would be possible to dynamically generate the
command line by looking at the devicetree state,
but this PR is a fixed implementation to simplify it.
This will be addressed separately.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
tests: drivers: entropy: api: add qemu_cortex_a53 configuration

Add `qemu_cortex_a53` target which has the `virtio,entropy` device.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
tests: drivers: build_all: virtio: add VIRTIO MMIO build test

Add the `qemu_cortex_a53` configuration to run the build test
for the VIRTIO MMIO driver.
To add this test, we have split the virtio-pci and virtio-mmio test cases.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

fkokosinski commented on drivers/virtio/virtio_mmio.c:
```diff
@@ -0,0 +1,419 @@
+
+	const uint32_t version = virtio_mmio_read32(dev, VIRTIO_MMIO_VERSION);
+
+	if (version < VIRTIO_MMIO_SUPPORTED_VERSION) {
```
```suggestion
	if (version != VIRTIO_MMIO_SUPPORTED_VERSION) {
```

It would probably be better to have this as an exact check in case new revision is introduced by the standard.

cfriedt replied:
Is this field broken up into semantic version numbers? If so, maybe the major version should be the deciding field. Typically minor versions should still be still compatible.

soburi replied:
@cfriedt 

This value is defined as single 32 bit number.
So, it can't take semver format.
(Now only defined 2 versions. 0x1 for legacy, 0x2 for current spec.)

@fkokosinski 

Thank you for your pointing out.
Fixed it.

fkokosinski commented on drivers/virtio/virtio_mmio.c:
```diff
@@ -0,0 +1,419 @@
+
+static void virtio_mmio_reset(const struct device *dev)
+{
+	virtio_mmio_write32(dev, VIRTIO_MMIO_STATUS, 0);
```
Probably none of the VMMs enter VM again before setting VIRTIO status to 0 (i.e. mid-reset), but the spec (2.4.2) states that reset should be considered complete when we read 0 from status, so it might be better to wait until reading status returns 0 before returning from this function.

fkokosinski commented on drivers/virtio/virtio_mmio.c:
```diff
@@ -0,0 +1,419 @@
+
+	k_spinlock_key_t key = k_spin_lock(&data->notify_lock);
+
+	virtio_mmio_write32(dev, VIRTIO_MMIO_QUEUE_NOTIFY, 0);
```
Shouldn't it be `virtio_mmio_write32(dev, VIRTIO_MMIO_QUEUE_NOTIFY, queue_idx)`? Because right now it's only able to notify the first queue.

fkokosinski commented on drivers/virtio/virtio_mmio.c:
```diff
@@ -0,0 +1,419 @@
+#define VIRTIO_MMIO_QUEUE_CONFIG_GENERATION 0x0fc
+#define VIRTIO_MMIO_CONFIG                  0x100
+
+#define DEVICE_STATUS_ACKNOWLEDGE 0
+#define DEVICE_STATUS_DRIVER      1
+#define DEVICE_STATUS_DRIVER_OK   2
+#define DEVICE_STATUS_FEATURES_OK 3
+#define DEVICE_STATUS_NEEDS_RESET 6
+#define DEVICE_STATUS_FAILED      7
+
+#define VIRTIO_F_VERSION_1 32
+
+/* Ranges of feature bits for specific device types (see spec 2.2)*/
+#define DEV_TYPE_FEAT_RANGE_0_BEGIN 0
+#define DEV_TYPE_FEAT_RANGE_0_END   23
+#define DEV_TYPE_FEAT_RANGE_1_BEGIN 50
+#define DEV_TYPE_FEAT_RANGE_1_END   127
+
+#define VIRTIO_MMIO_QUEUE_INTERRUPT                1
+#define VIRTIO_MMIO_DEVICE_CONFIGURATION_INTERRUPT 2
```
These defs are now provided by `virito_common.h`.

soburi replied:
Fixed to use common definition.

fkokosinski commented on drivers/virtio/virtio_mmio.c:
```diff
@@ -0,0 +1,419 @@
+	barrier_dmem_fence_full();
+}
+
+static void virtio_mmio_isr(const struct device *dev)
```
The bulk of this symbol can be replaced with `virtio_isr(dev, isr_status, data->virtqueue_count)` from `virito_common.c`.

soburi replied:
Ditto.

soburi committed:
drivers: entropy: virtio: Unify the notation to `VIRTIO`

Since `VirtIO` is not the official notation,
unify the name to `VIRTIO`.
In the text, `Virtio` and `virtio` can also be used
depending on the context.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi commented:
@fkokosinski @ceolin @cfriedt @kartben 

Continuing from https://github.com/zephyrproject-rtos/zephyr/pull/89439, could you please review this PR?

fkokosinski reviewed (approved):
VIRTIO MMIO LGTM

cfriedt reviewed (approved):
LGTM
